
**Scraping Toronto Neighborhoods Table from Wikipedia**

In [129]:
import pandas as pd

url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

df=pd.read_html(url, header=0)[0]

df.head()




Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [130]:
df.columns

Index(['Postal Code', 'Borough', 'Neighbourhood'], dtype='object')

In [131]:
df.rename(columns = {'Postal Code':'PostalCode'}, inplace = True) 
df.head()

PostalCode           Borough              Neighbourhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

clean and display the top 10 rows

In [132]:
#Data Cleaning 

df = df[df.Borough != 'Not assigned']
df = df[df.Borough!=0]
df.reset_index(drop=True, inplace=True)
i =0

for i in range(0, df.shape[0]):
  if df.iloc[i][2] == 'Not assigned':
    df.loc[i][2] = df.iloc[i][1]
    i = i+1
df = df.groupby(['PostalCode', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()
df.head()

PostalCode      Borough                           Neighbourhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [133]:
df = df.dropna()
empty = 'Not assigned'
df = df[(df.PostalCode != empty ) & (df.Borough != empty) & (df.Neighbourhood != empty)]

In [134]:
print('The dataframe has {} boroughs and {} neighbourhoods.'.format(
    len(df['Borough'].unique()),df.shape[0]))

The dataframe has 10 boroughs and 103 neighbourhoods.


In [135]:

def neighbourhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighbourhood'].tolist()))
                    
grp = df.groupby(['PostalCode', 'Borough'])
df2 = grp.apply(neighbourhood_list).reset_index(name='Neighbourhood')

In [136]:
print(df2.shape)
df2.head(11)

(103, 3)


PostalCode      Borough                                      Neighbourhood
0         M1B  Scarborough                                     Malvern, Rouge
1         M1C  Scarborough             Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough                  Guildwood, Morningside, West Hill
3         M1G  Scarborough                                             Woburn
4         M1H  Scarborough                                          Cedarbrae
5         M1J  Scarborough                                Scarborough Village
6         M1K  Scarborough        Kennedy Park, Ionview, East Birchmount Park
7         M1L  Scarborough                    Golden Mile, Clairlea, Oakridge
8         M1M  Scarborough    Cliffside, Cliffcrest, Scarborough Village West
9         M1N  Scarborough                        Birch Cliff, Cliffside West
10        M1P  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...

In [137]:
print('The DataFrame shape is', df2.shape)


The DataFrame shape is (103, 3)


The dataframe has 103 Postal codes but it has 212 rows, because each Postal code can present more than one neighborhood (210 in total). Therefore, the dataframe should be group by the Postal code, ending with a dataframe with 103 rows.

**I’m using “https://cocl.us/Geospatial_data” csv file to get all the geographical coordinates of the neighborhoods.**

In [138]:
import io 
import requests

url = 'https://cocl.us/Geospatial_data'

lat_long =requests.get(url).text
lat_long_df = pd.read_csv(io.StringIO(lat_long))
lat_long_df.head()


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [139]:
#rename columns
lat_long_df.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)


In [140]:
#merge tabels
df = pd.merge(df, lat_long_df, on = 'PostalCode')
df.head(11)

PostalCode      Borough  ...   Latitude  Longitude
0         M1B  Scarborough  ...  43.806686 -79.194353
1         M1C  Scarborough  ...  43.784535 -79.160497
2         M1E  Scarborough  ...  43.763573 -79.188711
3         M1G  Scarborough  ...  43.770992 -79.216917
4         M1H  Scarborough  ...  43.773136 -79.239476
5         M1J  Scarborough  ...  43.744734 -79.239476
6         M1K  Scarborough  ...  43.727929 -79.262029
7         M1L  Scarborough  ...  43.711112 -79.284577
8         M1M  Scarborough  ...  43.716316 -79.239476
9         M1N  Scarborough  ...  43.692657 -79.264848
10        M1P  Scarborough  ...  43.757410 -79.273304

[11 rows x 5 columns]

In [141]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


#### Use geopy library to get the latitude and longitude values of Toronto City.


In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.


In [142]:
address = 'Toronto City, T'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.


In [148]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df2['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in North York. So let's slice the original dataframe and create a new dataframe of the North York data.


In [163]:

North_York_data = df[df['Borough'] == 'North York'].reset_index(drop=True)
North_York_data.head()

PostalCode     Borough                 Neighbourhood   Latitude  Longitude
0        M2H  North York             Hillcrest Village  43.803762 -79.363452
1        M2J  North York  Fairview, Henry Farm, Oriole  43.778517 -79.346556
2        M2K  North York               Bayview Village  43.786947 -79.385975
3        M2L  North York      York Mills, Silver Hills  43.757490 -79.374714
4        M2M  North York       Willowdale, Newtonbrook  43.789053 -79.408493

In [164]:
address = 'North York, T'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of North York are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of North York are 43.7543263, -79.44911696639593.


As we did with all of New York City, let's visualizat North York the neighborhoods in it.


In [165]:
# create map of North York using latitude and longitude values
map_North_York = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(North_York_data['Latitude'], North_York_data['Longitude'], North_York_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_North_York)  
    
map_North_York

**Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.**

#### Define Foursquare Credentials and Version


In [166]:
CLIENT_ID = 'X0MHZA4D4RIBRBUQDASZLL1KBUBEFT5RDFA5FPXHPKWSUKCR' # your Foursquare ID
CLIENT_SECRET = '21P20TMNIHIBTPAVLEO15AIUZQK50MAYD1Z4H2YWZMS5LZ5F' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: X0MHZA4D4RIBRBUQDASZLL1KBUBEFT5RDFA5FPXHPKWSUKCR
CLIENT_SECRET:21P20TMNIHIBTPAVLEO15AIUZQK50MAYD1Z4H2YWZMS5LZ5F


#### Let's explore the first neighborhood in our dataframe.


Get the neighborhood's name.


In [168]:
North_York_data.loc[0, 'Neighbourhood']

'Hillcrest Village'

Get the neighborhood's latitude and longitude values.


In [170]:
neighborhood_latitude = North_York_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = North_York_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = North_York_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Hillcrest Village are 43.8037622, -79.3634517.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.


First, let's create the GET request URL. Name your URL **url**.


In [171]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
 # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=X0MHZA4D4RIBRBUQDASZLL1KBUBEFT5RDFA5FPXHPKWSUKCR&client_secret=21P20TMNIHIBTPAVLEO15AIUZQK50MAYD1Z4H2YWZMS5LZ5F&v=20180605&ll=43.8037622,-79.3634517&radius=500&limit=100'

Send the GET request and examine the resutls


In [172]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fa4269bc0877943bc013be8'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4ad9dce6f964a520651b21e3-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/golfcourse_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d1e6941735',
         'name': 'Golf Course',
         'pluralName': 'Golf Courses',
         'primary': True,
         'shortName': 'Golf Course'}],
       'id': '4ad9dce6f964a520651b21e3',
       'location': {'address': '10000 Dufferin Rd',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'distance': 197,
        'formattedAddress': ['10000 Dufferin Rd', 'Toronto ON', 'Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.805454826002794,
    

let's borrow the **get_category_type** function from the Foursquare lab.

In [173]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a _pandas_ dataframe.


In [174]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name                categories        lat  \
0               Eagle's Nest Golf Club               Golf Course  43.805455   
1                      AY Jackson Pool                      Pool  43.804515   
2                         Villa Madina  Mediterranean Restaurant  43.801685   
3                    Duncan Creek Park                   Dog Run  43.805539   
4  A.Y. Jackson Secondary School Track        Athletics & Sports  43.805068   

         lng  
0 -79.364186  
1 -79.366138  
2 -79.363938  
3 -79.360695  
4 -79.366677

And how many venues were returned by Foursquare?


In [175]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

5 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Manhattan


#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan


In [176]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called _manhattan_venues_.


In [178]:
# type your answer here
North_York_venues = getNearbyVenues(names=North_York_data['Neighbourhood'],
                                   latitudes=North_York_data['Latitude'],
                                   longitudes=North_York_data['Longitude']
                                  )

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Bedford Park, Lawrence Manor East
Lawrence Manor, Lawrence Heights
Glencairn
North Park, Maple Leaf Park, Upwood Park
Humber Summit
Humberlea, Emery


#### Let's check the size of the resulting dataframe


In [179]:
print(North_York_venues.shape)
North_York_venues.head()

(240, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Hillcrest Village              43.803762              -79.363452   
1  Hillcrest Village              43.803762              -79.363452   
2  Hillcrest Village              43.803762              -79.363452   
3  Hillcrest Village              43.803762              -79.363452   
4  Hillcrest Village              43.803762              -79.363452   

                                 Venue  Venue Latitude  Venue Longitude  \
0               Eagle's Nest Golf Club       43.805455       -79.364186   
1                      AY Jackson Pool       43.804515       -79.366138   
2                         Villa Madina       43.801685       -79.363938   
3                    Duncan Creek Park       43.805539       -79.360695   
4  A.Y. Jackson Secondary School Track       43.805068       -79.366677   

             Venue Category  
0               Golf Course  
1                      Pool  
2  Mediterranean Restaurant  
3                   Dog Run  
4        Athletics & Sports

Let's check how many venues were returned for each neighborhood


In [180]:
North_York_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
Don Mills                                                           23   
Downsview                                                           16   
Fairview, Henry Farm, Oriole                                        70   
Glencairn                                                            5   
Hillcrest Village                                                    5   
Humber Summit                                                        2   
Humberlea, Emery                                                     1   
Lawrence Manor, Lawrence Heights                                    13   
North Park, Maple Leaf Park, Upwood Park                             3   
Northwood Park, York University                                      7   
Parkwoods                                                            2   
Victoria Village                                                     5   
Willowdale, Willowdale East                                         33   
Willowdale, Willowdale West                                          5   
York Mills West                                                      2   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
Don Mills                                                            23   
Downsview                                                            16   
Fairview, Henry Farm, Oriole                                         70   
Glencairn                                                             5   
Hillcrest Village                                                     5   
Humber Summit                                                         2   
Humberlea, Emery                                                      1   
Lawrence Manor, Lawrence Heights                                     13   
North Park, Maple Leaf Park, Upwood Park                              3   
Northwood Park, York University                                       7   
Parkwoods                                                             2   
Victoria Village                                                      5   
Willowdale, Willowdale East                                          33   
Willowdale, Willowdale West                                           5   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
Don Mills                                           23              23   
Downsview                                           16              16   
Fairview, Henry Farm, Oriole                        70              70   
Glencairn                                            5               5   
Hillcrest Village                                    5               5   
Humber Summit                                        2               2   
Humberlea, Emery                                     1               1   
Lawrence Manor, Lawrence Heigh

#### Let's find out how many unique categories can be curated from all the returned venues


In [181]:
print('There are {} uniques categories.'.format(len(North_York_venues['Venue Category'].unique())))

There are 102 uniques categories.


## 3. Analyze Each Neighborhood


In [185]:
# one hot encoding
North_York_onehot = pd.get_dummies(North_York_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
North_York_onehot['Neighborhood'] = North_York_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [North_York_onehot.columns[-1]] + list(North_York_onehot.columns[:-1])
North_York_onehot = North_York_onehot[fixed_columns]

North_York_onehot.head(11)

Neighborhood  Accessories Store  Airport  \
0              Hillcrest Village                  0        0   
1              Hillcrest Village                  0        0   
2              Hillcrest Village                  0        0   
3              Hillcrest Village                  0        0   
4              Hillcrest Village                  0        0   
5   Fairview, Henry Farm, Oriole                  0        0   
6   Fairview, Henry Farm, Oriole                  0        0   
7   Fairview, Henry Farm, Oriole                  0        0   
8   Fairview, Henry Farm, Oriole                  0        0   
9   Fairview, Henry Farm, Oriole                  0        0   
10  Fairview, Henry Farm, Oriole                  0        0   

    American Restaurant  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0                     0            0                    0                 0   
1                     0            0                    0                 0   
2                     0            0                    0                 0   
3                     0            0                    0                 0   
4                     0            0                    0                 0   
5                     0            0                    0                 0   
6                     1            0                    0                 0   
7                     0            0                    0                 0   
8                     0            0                    0                 0   
9                     0            0                    0                 0   
10                    0            0                    0                 0   

    Athletics & Sports  Bakery  Bank  Bar  Baseball Field  Beer Store  \
0                    0       0     0    0               0           0   
1                    0       0     0    0               0           0   
2                    0       0     0    0               0           0   
3                    0       0     0    0               0           0   
4                    1       0     0    0               0           0   
5                    0       0     0    0               0           0   
6                    0       0     0    0               0           0   
7                    0       1     0    0               0           0   
8                    0       0     0    0               0           0   
9                    0       0     0    0               0           0   
10                   0       0     0    0               0           0   

    Bike Shop  Boutique  Bridal Shop  Bubble Tea Shop  Burger Joint  \
0           0         0            0                0             0   
1           0         0            0                0             0   
2           0         0            0                0             0   
3           0         0            0                0             0   
4           0         0            0                0             0   
5           0         0            0                0             0   
6           0         0            0                0             0   
7           0         0            0                0             0   
8           0         0            0                0             0   
9           0         0            0                0             0   
10          0         0            0                0             0   

    Burrito Place  Bus Station  Business Service  Butcher  Café  \
0               0            0                 0        0     0   
1               0            0                 0        0     0   
2               0            0                 0        0     0   
3               0            0                 0        0     0   
4               0            0                 0        0     0   
5               0            0                 0        0     0   
6               0            0                 0        0     0   
7               0            0      

And let's examine the new dataframe size.


In [186]:
North_York_onehot.shape

(240, 103)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [187]:
North_York_grouped = North_York_onehot.groupby('Neighborhood').mean().reset_index()
North_York_grouped

Neighborhood  Accessories Store  \
0   Bathurst Manor, Wilson Heights, Downsview North           0.000000   
1                                   Bayview Village           0.000000   
2                 Bedford Park, Lawrence Manor East           0.000000   
3                                         Don Mills           0.000000   
4                                         Downsview           0.000000   
5                      Fairview, Henry Farm, Oriole           0.000000   
6                                         Glencairn           0.000000   
7                                 Hillcrest Village           0.000000   
8                                     Humber Summit           0.000000   
9                                  Humberlea, Emery           0.000000   
10                 Lawrence Manor, Lawrence Heights           0.076923   
11         North Park, Maple Leaf Park, Upwood Park           0.000000   
12                  Northwood Park, York University           0.000000   
13                                        Parkwoods           0.000000   
14                                 Victoria Village           0.000000   
15                      Willowdale, Willowdale East           0.000000   
16                      Willowdale, Willowdale West           0.000000   
17                                  York Mills West           0.000000   
18                         York Mills, Silver Hills           0.000000   

    Airport  American Restaurant  Art Gallery  Arts & Crafts Store  \
0    0.0000             0.000000     0.000000             0.000000   
1    0.0000             0.000000     0.000000             0.000000   
2    0.0000             0.045455     0.000000             0.000000   
3    0.0000             0.000000     0.043478             0.000000   
4    0.0625             0.000000     0.000000             0.000000   
5    0.0000             0.014286     0.000000             0.000000   
6    0.0000             0.000000     0.000000             0.000000   
7    0.0000             0.000000     0.000000             0.000000   
8    0.0000             0.000000     0.000000             0.000000   
9    0.0000             0.000000     0.000000             0.000000   
10   0.0000             0.000000     0.000000             0.000000   
11   0.0000             0.000000     0.000000             0.000000   
12   0.0000             0.000000     0.000000             0.000000   
13   0.0000             0.000000     0.000000             0.000000   
14   0.0000             0.000000     0.000000             0.000000   
15   0.0000             0.000000     0.000000             0.030303   
16   0.0000             0.000000     0.000000             0.000000   
17   0.0000             0.000000     0.000000             0.000000   
18   0.0000             0.000000     0.000000             0.000000   

    Asian Restaurant  Athletics & Sports    Bakery      Bank       Bar  \
0           0.000000              0.0000  0.000000  0.095238  0.000000   
1           0.000000              0.0000  0.000000  0.250000  0.000000   
2           0.000000              0.0000  0.000000  0.000000  0.000000   
3           0.043478              0.0000  0.000000  0.000000  0.000000   
4           0.000000              0.0625  0.000000  0.062500  0.000000   
5           0.014286              0.0000  0.028571  0.028571  0.014286   
6           0.000000              0.0000  0.200000  0.000000  0.000000   
7           0.000000              0.2000  0.000000  0.000000  0.000000   
8           0.000000              0.0000  0.000000  0.000000  0.000000   
9           0.000000              0.0000  0.000000  0.000000  0.000000   
10          0.000000              0.0000  0.000000  0.000000  0.000000   
11          0.000000              0.0000  0.333333  0.000000  0.000000   
12          0.000000              0.0000  0.000000  0.000000  0.142857   
13          0.000000              0.0000  0.000000  0.000000  0.000000   
14          0.000000              0.0000  0.000

#### Let's confirm the new size


In [188]:
North_York_grouped.shape

(19, 103)

#### Let's print each neighborhood along with the top 5 most common venues


In [189]:
num_top_venues = 5

for hood in North_York_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = North_York_grouped[North_York_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor, Wilson Heights, Downsview North----
            venue  freq
0     Coffee Shop  0.10
1            Bank  0.10
2  Ice Cream Shop  0.05
3     Bridal Shop  0.05
4   Deli / Bodega  0.05


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Café  0.25
2                 Bank  0.25
3  Japanese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0      Sandwich Place  0.09
1  Italian Restaurant  0.09
2         Coffee Shop  0.09
3       Grocery Store  0.05
4    Greek Restaurant  0.05


----Don Mills----
                 venue  freq
0                  Gym  0.13
1          Coffee Shop  0.09
2  Japanese Restaurant  0.09
3           Beer Store  0.09
4   Dim Sum Restaurant  0.04


----Downsview----
            venue  freq
0   Grocery Store  0.19
1            Park  0.12
2  Baseball Field  0.06
3           Hotel  0.06
4         Airport  0.06


----Fairview, Henry Farm, Or

#### Let's put that into a _pandas_ dataframe


First, let's write a function to sort the venues in descending order.


In [190]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [191]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = North_York_grouped['Neighborhood']

for ind in np.arange(North_York_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(North_York_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
1                                  Bayview Village    Chinese Restaurant   
2                Bedford Park, Lawrence Manor East        Sandwich Place   
3                                        Don Mills                   Gym   
4                                        Downsview         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Bank     Mobile Phone Shop        Sandwich Place   
1                  Café                  Bank   Japanese Restaurant   
2    Italian Restaurant           Coffee Shop      Greek Restaurant   
3            Beer Store           Coffee Shop   Japanese Restaurant   
4                  Park                  Bank          Liquor Store   

  5th Most Common Venue 6th Most Common Venue       7th Most Common Venue  \
0           Gas Station                 Diner              Ice Cream Shop   
1         Women's Store        Discount Store  Construction & Landscaping   
2         Grocery Store     Indian Restaurant                   Juice Bar   
3  Caribbean Restaurant        Clothing Store          Chinese Restaurant   
4                 Hotel         Shopping Mall                Home Service   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0         Deli / Bodega    Chinese Restaurant  Middle Eastern Restaurant  
1     Convenience Store        Cosmetics Shop              Deli / Bodega  
2          Liquor Store             Locksmith    Comfort Food Restaurant  
3                  Café    Dim Sum Restaurant         Italian Restaurant  
4        Baseball Field      Business Service       Gym / Fitness Center

## 4. Cluster Neighborhoods


Run _k_-means to cluster the neighborhood into 5 clusters.


In [201]:
# set number of clusters
kclusters = 5

North_York_grouped_clustering = North_York_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(North_York_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 4, 4, 4, 4, 4, 4, 3, 2], dtype=int32)